enhancement(datadog source): add API key validation for Datadog logs source

ChinYing-Li opened the pull request:
<!--
**Your PR title must conform to the conventional commit spec!**

  <type>!?(<scope>): <description>

  * `type` = chore, enhancement, feat, fix
  * `!` = signals a breaking change
  * `scope` = https://github.com/timberio/vector/blob/master/.github/semantic.yml#L4
  * `description` = short description of the change

Examples:

  * enhancement(file source): Added `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fixed a bug discovering new files
  * chore(external docs): Clarified `batch_size` option
-->
This PR aims to solve #6809. Any suggestion is appreciated!

prognant reviewed (commented):
Thanks a lot for your contribution !

I like the idea of being able to whitelist some API keys. Two things, as this PR aims to introduce new configuration flags the doc should be updated accordingly in `docs/reference/components/sources/datadog_logs.cue`. Also I was thinking that introducing a boolean settings like `drop_invalid_api_key` that, if set true, would cause vector to drop logs coming with an API not listed in the `valid_api_keys` list, WDYT ?

ChinYing-Li commented:
> Thanks a lot for your contribution !
> 
> Also I was thinking that introducing a boolean settings like `drop_invalid_api_key` that, if set true, would cause vector to drop logs coming with an API not listed in the `valid_api_keys` list
Added the "drop_invalid_api_key" flag. PR ready for review, thank you!

prognant reviewed (commented):
Regarding the dropping part as detail in #6809 I think we should opt for the following behaviour:
* If the drop setting is `true` then drop event coming with an unknown API key
* If the drop setting is `false` then just ignore the API key and don't record it with the event

ChinYing-Li commented:
Sorry for coming back late! I have added a few extra test cases, and in the unit test `drop_event_with_invalid_api_key`  where two events are sent, I expect the first event to be dropped. While the first event is indeed, dropped, the second event's `dd_api_key` is lost:
```bash
---- sources::datadog::logs::tests::drop_event_with_invalid_api_key stdout ----
thread 'sources::datadog::logs::tests::drop_event_with_invalid_api_key' panicked at 'Key is not found: "dd_api_key"', src/sources/datadog/logs.rs:325:24

```
I wonder if I am missing something?

jszwedko commented:
Thanks for the note @ChinYing-Li ! @prognant do you mind taking another look at this?

spencergilbert commented:
Also the new CLA requirement 😄

prognant commented:
Indeed the code has been significantly rewritten since your original PR, nevertheless the feature you suggested remains relevant if you have some times to rebase your PR against the current code

ChinYing-Li commented on src/sources/datadog/agent.rs:
```diff
@@ -196,6 +216,13 @@ impl DatadogAgentSource {
             .or_else(|| header.map(Arc::from))
     }
 
+    fn is_valid_api_key(&self, api_key: &Option<Arc<str>>) -> bool {
+        if let (Some(k), Some(valid_api_keys)) = (api_key, &self.valid_api_keys) {
+            return valid_api_keys.contains(&k.to_string());
+        }
+        true
+    }
+
```
Here's the function for validating the api_key. Does it make sense to accept any events if `self.valid_api_keys` is `None`? Also, if `self.valid_api_keys` is an empty vector and `self.drop_api_key` is true, all events will rejected.

ChinYing-Li replied:
@prognant Just a friendly ping to see if you have any suggestion. Thanks!

prognant commented:
Sorry for the late answer, I think we should stick to the behaviour described in #6809 and keep the boolean settings to either accept only event with a valid api key or remove unknown api keys.

Anyway the `datadog_agent` source now supports metrics & traces on top of logs, we would need all type of event covered, this would likely imply modification [here](https://github.com/vectordotdev/vector/blob/c61794c/src/sources/datadog/agent/mod.rs#L223-L248), the `ApiKeyExtractor` struct would then hold a list of valid api key and the `extract` method could could return an extra boolean to figure out whether the api key found is in the valid list. 

But as already said this would be a nice addition.

jszwedko commented:
Closing as stale, but we'd be happy to see this work picked back up in this branch or a new one.
